Sally7/S7Connection.cs

```diff
@@ -104,58 +100,98 @@ public void Dispose()
-            S7ConnectionHelpers.ParseConnectionConfirm(buffer.AsSpan().Slice(0, length));
+                using var closeOnCancel = linkedToken.MaybeUnsafeRegister(TcpClient.Close);
 
-#if NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
-            await stream.WriteAsync(buffer, 0, S7ConnectionHelpers.BuildCommunicationSetup(buffer), cancellationToken).ConfigureAwait(false);
+#if NET5_0_OR_GREATER
+                await TcpClient.ConnectAsync(host, IsoOverTcpPort, linkedToken).ConfigureAwait(false);
 #else
-            cancellationToken.ThrowIfCancellationRequested();
-            await stream.WriteAsync(buffer, 0, S7ConnectionHelpers.BuildCommunicationSetup(buffer)).ConfigureAwait(false);
+                linkedToken.ThrowIfCancellationRequested();
+                await TcpClient.ConnectAsync(host, IsoOverTcpPort).ConfigureAwait(false);
 #endif
-            length = await ReadTpktAsync(stream, buffer, cancellationToken).ConfigureAwait(false);
-            S7ConnectionHelpers.ParseCommunicationSetup(buffer.AsSpan().Slice(0, length), out var pduSize, out var maxRequests);
 
-            Parameters = new S7ConnectionParameters(pduSize, maxRequests);
-            bufferSize = Parameters.GetRequiredBufferSize();
+                var stream = TcpClient.GetStream();
+
+                var buffer = ArrayPool<byte>.Shared.Rent(100);
+                try
+                {
+                    await stream.FrameworkSpecificWriteAsync(buffer, 0,
+                            S7ConnectionHelpers.BuildConnectRequest(buffer, sourceTsap, destinationTsap),
+                            cancellationToken)
+                        .ConfigureAwait(false);
 
-            memoryPool ??= new Sally7MemoryPool(bufferSize);
+                    var length = await ReadTpktAsync(stream, buffer, linkedToken).ConfigureAwait(false);
+                    S7ConnectionHelpers.ParseConnectionConfirm(buffer.AsSpan().Slice(0, length));
 
-            requestExecutor = executorFactory.Invoke(this);
+                    await stream
+                        .FrameworkSpecificWriteAsync(buffer, 0, S7ConnectionHelpers.BuildCommunicationSetup(buffer), linkedToken)
+                        .ConfigureAwait(false);
 
-            ArrayPool<byte>.Shared.Return(buffer);
+                    length = await ReadTpktAsync(stream, buffer, linkedToken).ConfigureAwait(false);
+                    S7ConnectionHelpers.ParseCommunicationSetup(buffer.AsSpan().Slice(0, length), out var pduSize,
+                        out var maxRequests);
+
+                    Parameters = new S7ConnectionParameters(pduSize, maxRequests);
+                    bufferSize = Parameters.GetRequiredBufferSize();
+
+                    memoryPool ??= new Sally7MemoryPool(bufferSize);
+
+                    requestExecutor = executorFactory.Invoke(this);
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
Too bad there isn't a IDisposable helper to take care of this...

I really don't miss the days where it is so hard to write a simple destructor allowing for powerful RAII classes, or quickly use a ScopeGuard with a lambda as in C++ :-)
(But of course other things like async and LINQ are sorely missed)

Reply:
I'm not sure why they chose this pattern, but creating a disposable pattern for this would be easy-peasy of course. Perhaps the disposable pattern has more of an performance impact, or is often not applicable because the buffer lives longer than the method creating it.

Reply:
> often not applicable because the buffer lives longer than the method creating it

Yep, that's the reason -- especially together with the default implementation of the MemoryPool which is / was a motivation to have this type.